### PR TITLE
Data Interview Post

### PR DESCRIPTION
**Pull request by agude**
Still need to fill in notebooks and links, but I think the prose is good to go!

### CODE REVIEW

**Comment by charles-uno** [comment]
How many posts do you have total? Would it make more sense to show the lead image by default, and have the option be `hide_lead_image`?

File: _layouts/post.html [hunk 1]
```diff
@@ -24,6 +24,13 @@ <h1 class="post-title">{{ page.title }}</h1>
     </p>
   {% endif %}
 
+  {% if page.show_lead_image and page.image %}
+    <img
```
Comment (charles-uno) [inline _layouts/post.html:28]: Should there be a class or id on here?

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 2]
```diff
@@ -0,0 +1,370 @@
+---
+layout: post
+title: "Data Science Interview Practice: Data Manipulation"
+description: >
+image: /files/switrs-dataset/tram_auto_crash_in_1957_frederiksplein_amsterdam.jpg
+show_lead_image: True
```
Comment (charles-uno) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:6]: X

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+image_alt: >
+categories:
+  - career-advice
+  - interview-prep
+---
+
+<!-- Simple script to call from the "Show/Hide" buttons-->
+<script type="text/javascript">
+function showhide(d) {
+  d.style.display = (d.style.display !== "none") ? "none" : "block";
```
Comment (charles-uno) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:16]: Prefer `elt` to `d`. Ternary operator unnecessarily trades legibility for brevity.

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+}
+</script>
+
+{% assign twitter-name = site.author.twitter %}
```
Comment (charles-uno) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:20]: This seems like it should be in the header or template or something. Do you have to spell it out in every post?

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+
+{% capture file_dir %}/files/interview-prep{% endcapture %}
+
+I often get asked by newly-minted PhDs trying to get their first data job:
+
+> How can I prepare for dataset-based interviews? Do you have any examples of
+> datasets to practice with?
+
+I never had a good answer. I would tell them about how the interviews worked,
+but I wished I had something to share that they could work with and practice
```
Comment (charles-uno) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:30]: "work with and practice on" -> "get their hands on"?

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md (hunk 4 above)
Comment (agude) [reply _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:19]: Actually I should just use site.author.twitter... No reason to alias it.

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+
+I never had a good answer. I would tell them about how the interviews worked,
+but I wished I had something to share that they could get their hands on.
+
+As of today, that's changing. In this post I put together a series of practice
+questions like the kind you might see (or be expected to come up with) in a
+hands-on data interview using the [curated and hosted dataset of California
+Traffic accidents][switrs_dataset]. The dataset is available for download from
+both [Kaggle][kaggle] and [Zenodo][zenodo], and I even have an [example
+notebook][example_notebook] for how to work with the data entirely online
```
Comment (Veldrina) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:42]:
```suggestion
notebook][example_notebook] demonstrating how to work with the data entirely online
```

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+within Kaggle.
+
+[switrs_dataset]: {% post_url 2020-11-24-switrs_sqlite_hosted_dataset %}
+[kaggle]: https://www.kaggle.com/alexgude/california-traffic-collision-data-from-switrs
+[zenodo]: https://zenodo.org/record/4284843
+[example_notebook]: https://www.kaggle.com/alexgude/starter-california-traffic-collisions-from-switrs
+
+## Interview Format
+
+As I mentioned in [my post about my last interview experience][last_post],
```
Comment (Veldrina) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:52]:
```suggestion
As I mentioned in [my post about my most recent interview experience][last_post],
```

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+data science and machine learning interviews have gotten more practical,
```
Comment (Veldrina) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:53]:
```suggestion
data science and machine learning interviews have become more practical,
```

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+|------------------:|
+|         9,172,565 |
+
+</div>
+
+</div>
+
+### What percent of collisions involve males aged 16--25?
+
+Young men are famously unsafe drivers, lets look at how many collisions
```
Comment (Veldrina) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:108]:
```suggestion
Young men are famously unsafe drivers so let's look at how many collisions
```

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+they're involved in.
+
+<button id="button" onclick="showhide(hidden3)">Show solution</button>
+<div class="hidden" id="hidden3" markdown="1" style="display: none;">
+
+The age and gender of the drivers are in the parties table so the query does a
+simple filter on those entries. The tricky part comes from needing the ratio:
+we have to get the total number of collisions. We could hard-code it, but I
```
Comment (Veldrina) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:116]:
```suggestion
simple filter on those entries. The tricky part comes from needing to calculate the ratio as this
requires us to get the total number of collisions. We could hard-code the lookup, but I
```

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+|        0.242 |
+
+</div>
+
+</div>
+
+### How many solo motorcycle crashes are there per year?
+
+A "_solo_" crash is one where the driver runs off the road or hits a
+stationary object. How many solo motorcycle crashes where there each year? Why
```
Comment (Veldrina) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:146]:
```suggestion
stationary object. How many solo motorcycle crashes were there each year? Why
```

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+is 2020 so low?
```
Comment (Veldrina) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:147]:
```suggestion
does 2020 seem to (relatively) have so few?
```

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 13]
```diff
@@ -0,0 +1,378 @@
...
+|             2016 |              4471 |
+|             2017 |              4373 |
+|             2018 |              4240 |
+|             2019 |              3772 |
+|             2020 |              2984 |
+
+</div>
+
+The count is low in 2020 primarily because the data doesn't cover the whole
+year. It is also low due to the COVID pandemic which kept people off the
```
Comment (Veldrina) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:199]:
```suggestion
year. It is also low due to the COVID pandemic keeping people off the
```

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+
+</div>
+
+### What make of vehicle has the largest fraction of accidents on the weekend? During the work week?
+
+Weekdays are generally commute and work-related traffic, while weekends
+involves recreational travel. Do we see different vehicles involved in
+collisions on these days?
+
+Only consider vehicle makes with at least 10,000 collisions.
```
Comment (Veldrina) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:211]: Consider expanding this sentence fragment.

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+
+<button id="button" onclick="showhide(hidden4)">Show solution</button>
+<div class="hidden" id="hidden4" markdown="1" style="display: none;">
+
+This query is tricky. We need to aggregate collisions by vehicle make, which
+means we need the parties table. We also care about when the crash happened,
+which means we need the collisions table. So we need to join these tables
```
Comment (Veldrina) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:218]:
```suggestion
which means we need the collisions table. So we need to join these two tables
```

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 16]
```diff
@@ -0,0 +1,378 @@
...
+together.
+
+I use a sub-query to do the aggregation. A `WTIH` clause keeps it tidy so we
+don't have to copy/paste the sub-query twice. I use `HAVING` to filter out
+makes with too few collisions; it has to be `HAVING` and not `WHERE` because
+it filters **after** the aggregation.
+
+I then construct two queries that read from the sub-query to select the
+highest row for the weekend and weekdays. I `UNION` to two queries together so
```
Comment (Veldrina) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:227]:
```suggestion
highest row for the weekend and weekdays. I `UNION` the two queries together so
```

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 17]
```diff
@@ -0,0 +1,378 @@
...
+we end up with a single table containing our results. The double select is to
+allow the `ORDER BY` before the `UNION`.
+
+In an interview setting, I would have just written two simpler queries: one
+that gets the highest weekend fraction and one that gets the highest weekday
+fraction with a lot of copy and pasted code. This would have been a lot faster
+to come up with and write.
```
Comment (Veldrina) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:234]: Would it make sense to show these two simpler queries? Maybe discuss how one can answer an interview using the two simple queries then discuss how they can be combined into a single query? (are there advantages or drawbacks to that approach?)

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 18]
```diff
@@ -0,0 +1,378 @@
...
+These results makes sense, Peterbilt is a commercial truck manufacturer which
+you expect to be driven for work. Harley-Davidson makes iconic motorcycles
+that people ride for fun on the weekend with their friends.
+
+</div>
+
+### How many different values represent "Toyota" in the Parties database? How would you go about correcting for this?
+
+Data is **_never_** as clean as you would hope, even after I [curated it for
+you][switrs_dataset]. How many different ways does "Toyota" show up?
```
Comment (Veldrina) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:296]:
```suggestion
Data is **_never_** as clean as you would hope,  and this applies even to the [curated
SWITRS dataset][switrs_dataset]. How many different ways does "Toyota" show up?
```

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 19]
```diff
@@ -0,0 +1,378 @@
...
+| TOYTT          |             6 |
+| TOYOY          |             6 |
+| TOYOTS         |             5 |
+| TYOTA          |             4 |
+| ...            |           ... |
+
+</div>
+
+Most of those look like they mean Toyota, although Tymco is a different
+company the makes street sweepers.
```
Comment (Veldrina) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:357]:
```suggestion
company that makes street sweepers.
```

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md [hunk 20]
```diff
@@ -0,0 +1,378 @@
...
+
+Here is how I would handle this issue: the top 5 make up the vast majority of entries.
+I would fix those by hand and move on. More generally it seems that makes are
+represented mostly by their name or a four-letter abbreviation. It wouldn't
+be too hard to fix these for the most common makes.
```
Comment (Veldrina) [inline _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:362]:
```suggestion
be too hard to detect and fix these for the most common makes.
```

File: _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md (hunk 17 above)
Comment (agude) [reply _drafts/2020-12-14-data_science_interview_prep_data_manipulation.md:234]: Great feedback!

**Pull request merged**